Explore more monorepo use-cases

csomh opened the pull request:
The original monorepo research was made for source-git. Extend this by
looking at how this can be used in upstream projects with Packit.

Signed-off-by: Hunor Csomortáni <csomh@redhat.com>

csomh commented:
Okay, I just realized, that maybe `packages` should be a dictionary instead of a list, as we would expect that each item in `packages` is a different package, with a different name.

If the schema is changed like this, then the package name could be used as a default for `downstream_package_name`.

Examples:

```yaml
packages:
  backend:
    downstream_package_name: copr-backend
    ...
  frontend:
    downstream_package_name: copr-frontend
    ...

# When referencing a package, package name (the key from the 'packages' dict) should be used
jobs:
- job: copr_build
  trigger: pull_request
  packages:
  - backend
  - frontend
```

The above could be written without `downstream_package_name`:

```yaml
packages:
  copr-backend:
    ...
  copr-frontend:
    ...

# When referencing a package, package name (the key from the 'packages' dict) should be used
jobs:
- job: copr_build
  trigger: pull_request
  packages:
  - copr-backend
  - copr-frontend
```

@lachmanfrantisek @TomasTomecek WDYT? Should I change the proposal like this?

jpopelka reviewed (commented):
looks OK so far, +1 to change `packages` to a dictionary

TomasTomecek commented on source-git-monorepos/README.md:
```diff
@@ -1,41 +1,80 @@
-repository. They are a special use-case from a source-git point of view,
-because an upstream monorepo maps to multiple dist-git repositories in the
-distribution.
+Monorepos store the source code of multiple projects in a single repository.
```
I wonder if want to account for different build configurations too, e.g. have different compiler flags for multiple sets of builds.

csomh replied:
I would expect to find things related to build configuration in the spec-file of each package (or some other scripting called during *build-time*), and from this point of view I would say these fall outside the scope of source-git. Or am I missing something?

TomasTomecek replied:
I am thinking out loud here as I see a big potential in the monorepo concept that it could solve more use cases related to building.

rpmbuild and mock [support](https://github.com/rpm-software-management/mock/blob/main/mock/docs/mock.1#L417) setting [conditionals build options](https://rpm-software-management.github.io/rpm/manual/conditionalbuilds.html). AFAIK no one requested this functionality from us yet (not used anymore?). Nevertheless, I just wanted to mention the feature here.

No need to change anything in the research doc. Let's keep the scope down.

mfocko replied:
I'd be more afraid about shared sources between packages, e.g. helpers, utility libraries.

_edit_: never mind, there's `package_paths`

TomasTomecek commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,355 @@ packages:
 `.distro/source-git.yaml`. Downstream files to be synced to dist-git are to be
-kept in subdirectories, named according to the distro package, in `.distro` to
+kept in subdirectories, named according to `package_name`, in `.distro`, to
 keep things organized.
 
 This way all distribution configs and files would be kept at one place, which
-probably is somewhat better than having source-git configuration spread
-across the monorepo, in subdirectories. This setup also allows multiple paths
-from the repo mapped to a dist-git repo.
+probably is somewhat better than having source-git configuration spread across
+the monorepo, in subdirectories. This setup also allows multiple paths from
+the repo mapped to a dist-git repo.
```
I'm confused here: so where would I find a specfile?

csomh replied:
By convention in source-git specfiles would be stored in `.distro/<package_name>/<dist_git_name>.spec`. I've added an example above to make this more clear.

In a Packit context, it doesn't really matter, as everything is configurable.

TomasTomecek replied:
That makes sense, I like that the spec files are in a single place. I thought about how archives for these packages are created - but Packit can already work with setups where spec files and sources are in different directories.

TomasTomecek commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,355 @@ packages:
+
+### Internal Design
+
+With monorepos needs to be taught to map a single upstream repository
+(`PackitAPI.up`) to multiple dist-git/downstream repositories. This means that
+`PackitAPI.dg` should become a dictionary holding the objects for each
+dist-git repository.
+
+In order to make the transition easier, we could keep `.dg` as it is, and make
+it to return the dist-git repo for the first (and only) package. If there are
+more packages defined, an exception should be raised. This way parts of the
+API and CLI which were not yet updated to support monorepos, could continue to
+work, but would fail if used with a monorepo configuration specifying multiple
+packages.
```
Oh my, this will indeed be a lot of work

TomasTomecek reviewed (commented):
This looks like a great hackfest material.

mfocko commented on source-git-monorepos/README.md:
```diff
@@ -1,41 +1,80 @@
+When processing a package from a monorepo, Packit should operate within these
+paths, that is, limit patch generation from Git-history to these paths.
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
```
(definitely can't tell the editor :eyes:)

csomh replied:
:see_no_evil:

mfocko commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,369 @@ packages:
+
+### packit prepare-sources
+
+I'm not entirely sure my understanding of what this command does is correct 😅
```
It's run as a preparation step when building SRPMs in Copr.

mfocko commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,369 @@ packages:
+package in the `packages` section.
+
+The example above could be made more useful if extending lists would be
+supported by Packit. (YAML doesn't support this, one gets nested lists when
```
fmf… but that would be another story I guess

csomh replied:
Yes, it would another story. This could be even done separately from this whole monorepo-saga. Stackowerflow suggested something like:

```yaml
list1: &listref
- element1
- element2
list2: !flatten
- *listref
- additional-element
```

We would need to write the code to flatten `[['element1', 'element2'], 'additional-element']`, and set a trigger to do it when the `!flatten` tag is met.

lachmanfrantisek commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+
+### Impact on Actions
+
+Because actions can now be defined on the top-level, package-level and
+job-level, one might ask, in what path these should be executed.
+
+Because a package can map to multiple sub-paths in the monorepo, the most
+sensible thing to do is to keep it simple and execute each action at the
+top-level, even if the action is defined on a job- or package-level.
```
Yes, that was my concern in the first question. But if we preserve current use-cases and document the behaviour of monorepo users, I don't see a problem with choosing any approach.

lachmanfrantisek replied:
:+1: Exactly.

lachmanfrantisek reviewed (commented):
Wow, that's really detailed!

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+- downstream_package_name: copr-backend
+  package_path: backend
+  dist_git_path: rpms/copr-backend
+  dist_git_branch: rawhide
```
I tend to feel that this should be a list of branch names?

csomh replied:
Why so?

`source-git init` initializes a single branch in a source-git repo. Furthermore: source-git branches map to dist-git branches one-to-one.

praiskup replied:
Hm, I guess? I'm offtopic (I thought it is the shared config format with the packit service, and the "propose downstream" configuration).

Just from the experience with tito... where you configure the precise set of branches you want to automatically update with new releases (aka `tito release`).

csomh replied:
This configuration is a proposal that could make **initializing** source-git monorepos from dist-git repositories easier. It's not planned to be used as a source-git config per se.

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+
+The command should check the sync status of a sub-package from a source-git
+monorepo and the corresponding dist-git repo. `downstream_package_name` should
+match `basename DIST_DIS`.
```
DIST_DIS?

csomh replied:
Thanks!

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+- downstream_package_name: copr-cli
+  package_path: cli
+  dist_git_path: rpms/copr-cli
+  dist_git_branch: rawhide
```
I bet only Fedora dist-git is supported, now or forever?  Copr identifies also CentOS, CentOS Stream and Red Hat (internally). Plus four Copr dist-git instances, but these are not open for PRs.  So just a feedback, DistGit operations in Copr have the `--distgit <DISTGIT_ID` option (which hides the dist-git instance configuration details).

csomh replied:
All dist-gits are supported, I just tend to use Fedora in examples.

As of now mapping the source-git repo to a dist-git instance is not configurable for the users only for the bots, as we expect CentOS to want to have their own source-git repos in their own place (which are mapped to CentOS dist-git) and Fedora to want their own source-git repos in their own place, mapped to Fedora dist-git repos.

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+**Going further**, when used with monorepos, this command could be able to
+update multiple dist-git repositories. This means, that the `DIST_GIT`
+argument of the command would become `DIST_GIT [DIST_GIT...]` (one or more
+paths to dist-git repositories).
```
In Tito, you do `tito release` from within the upstream sub-directory.  E.g. you
do `cd backend && tito release <releaser_id>`.  Not sure what is better, but just
noting that it might be worth considering.

csomh replied:
Thanks for mentioning this.

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+On the other hand, the command needs to be updated to support a source-git
+repo as a target when the source-git repo is a monorepo. In this case only the
+paths of the sub-package need to be considered during the update. `DIST_GIT`
+should match one of the `downstream_package_name`s in `.distro/source-git.yaml`.
```
This command could IMO be called also from within the source-git subdirectory.

csomh replied:
This command could be called from any path as it'll receive 3 arguments: the path to the dist-git repository, the path to the source-git repo, and a revision range in dist-git.

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+### Hardly
+
+When opening an MR in a source-git monorepo, Hardly should figure out which
+dist-git packages are going to be impacted by the changed (by looking at
```
s/changed/change/

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+  copr-backend:
+    specfile_path: backend/copr-backend.spec
+    package_paths:
+      - backend
```
In Copr we by default search for `<package_path>/<package_name>.spec`, to simplify the config.

csomh replied:
I've made a proposal for this: see the "Package and specfile discovery" chapter.

praiskup commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+### packit srpm
+
+The command should receive a new optional argument, `[PACKAGE...]`, to define
+zero or more packages for which an SRPM should be built.
```
Ditto here, having a possibility to call this also from within sub-dir would be plus.

csomh replied:
Although plain `packit` commands prefer to guess their arguments from their context, context might be ambiguous so I try to follow the "explicit is better than implicit" mantra to avoid confusion.

If the current proposal is going to be followed than calling `packit srpm` from a sub-dir would result in building an SRPM for all the packages in the monorepo (Packit would look up the config in the repo-root, and b/c there is not `PACKAGE` argument provided, it will build an SRPM for all the packages defined in the config.

Of course, we could say, that if the current sub-dir is a `package_path` for a package in the config, then build just that SRPM. Which is fine, if there is only one `package_path` in a package, but becomes more complicated if there are two or more. And to avoid this complication and possible user confusion, I would prefer to stay explicit :)

csomh replied:
The only example I know is in LLVM. For their daily builds, the do something source-git-like, and for `clang` they [fetch 2 paths](https://github.com/kwk/llvm-project/blob/0eacd884dd720b7a2156fb84df910388c5677b33/clang/clang.spec#L88) from their monorepo.

csomh commented on source-git-monorepos/README.md:
```diff
@@ -1,41 +1,80 @@
+`{package_name: package_object}`. Package-objects whould have keys identical
+to the current top-level keys in `.packit.yaml` and `source-git.yaml`, except
+`upstream_project_url` and `upstream_ref`—these are left to be top-level keys
+only, and a new `package_paths` key which is a list of paths in the monorepo
```
> I have a few questions about `package_paths`:
> 
> * Is this the opposite of [RFE: Excluded paths packit-service#545](https://github.com/packit/packit-service/issues/545) (=react only on changes in this path(s))? Or use it just for source manipulation like archive/path generation? Or, a working directory for our jobs? Or all?
>   
>   * Maybe, we can introduce an option for each usecase -- `source_path`, `include_path`/`exclude_path`, `working_path`.

That's a good question.

I see `package_paths` as only used for archive and patch generation. The _Excluded paths_ RFE is different in the sense that it tells when to run or not to run Packit.

`package_paths` might also be used for this purpose, but that might be more problematic: packages in the monorepo might be interdependent, and some projects would like to build all of them on every change, other projects just the packages which had code changed. So I would see that as a separate config key.

As for `working_path`, I think it makes sense to keep that at the root of the repository, to not to complicate things.

> * Will this be required? (I can think of some use-cases without having distinct subdirectories for each package.)

I would be curious to see some examples for this. So far the monorepos I've seen (llvm, copr) have generated packages from one or more sub-directories in the directory.

csomh replied:
I took some of your suggestions and made a proposal to have `handle.paths`, in order to be able to fine tune what paths services should react to (or not), independent of `package.paths`. See the "`package.paths` and subdirectories to react on" chapter.

I'm still not sure about working directories, but I've took note of this uncertainty and made a separate chapter for this, too.  Of course we can discuss this further if you want: I'm open to be convinced that this is essential to have a good monorepo support.

lachmanfrantisek replied:
> I'm open to be convinced that this is essential to have a good monorepo support.

No, that's maybe more an enhancement.

> I would be curious to see some examples for this. So far the monorepos I've seen (llvm, copr) have generated packages from one or more sub-directories in the directory.

I know it's not a true monorepo, but e.g. having multiple different spec-files using the same source tree -- e.g. `/fedora/some.spec` and `/rhel/some.spec`, but source in `/project`.

csomh commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
-probably is somewhat better than having source-git configuration spread
-across the monorepo, in subdirectories. This setup also allows multiple paths
-from the repo mapped to a dist-git repo.
+probably is somewhat better than having source-git configuration spread across
```
This sentence is in the source-git section, so I'll probably not apply this suggestion :)

In source-git I think it makes sense to have a more opinionated layout of the repository in order to be consistent across packages in a distribution. For Packit, I don't know if this is important. Upstream projects using Packit might choose to put files to different places.

One thing we should avoid is to encourage projects using Packit to create a `.distro` directory upstream, as that would conflict with the source-git layout :D

csomh commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
-from the repo mapped to a dist-git repo.
+probably is somewhat better than having source-git configuration spread across
+the monorepo, in subdirectories. This setup also allows multiple paths from
+the repo mapped to a dist-git repo.
```
Added "single" to be more explicit.

csomh commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+  copr-frontend:
+    specfile_path: frontend/copr-frontend.spec
+    package_paths:
+      - frontend
```
Actually it does not mean that. `package_paths` only maps monorepo paths to dist-git repositories. But it doesn't tell when actions using that mapping should be done. I think I'll need to revisit this topic once again and be more detailed about the intent, b/c @lachmanfrantisek brought up a similar topic.

praiskup replied:
Yeah, I just feel mandatory to saying this aloud :-) This is an absolute must for monorepos like Copr is, we simply couldn't afford to rebuild all the large set packages for each PR.  On top of this, it would be nice to have a way to define "order" for building them, in case multiple of them are affected by a single PR and there are dependencies between them.

csomh replied:
Added a new chapter "`package.paths` and subdirectories to react on", which makes a proposal to deal with this.

csomh commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+
+### Internal Design
+
+With monorepos needs to be taught to map a single upstream repository
+(`PackitAPI.up`) to multiple dist-git/downstream repositories. This means that
+`PackitAPI.dg` should become a dictionary holding the objects for each
+dist-git repository.
```
Wouldn't this approach leak logic regarding the interpretation of the configuration to the CLI or the service and potentially cause re-implementing similar logic in both places?

lachmanfrantisek replied:
I don't think it's a huge issue. Maybe, we can share the common code at least. This is the first approach I would try but maybe we will find some issues when working on this for real.

csomh commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
+    targets:
+      - fedora-all
+    packages:
+      - copr-cli
```
Obviously this is just and example for how the configuration should look like, which by chance it uses `copr`, but in no way should this be considered as a reference config for `copr`.

praiskup replied:
Since nobody except for us will use/read this example, I think it is fine.  But if you wanted to use this as an example in docs, I would prefer switching it (clients supported everywhere, servers only on Fedora, as more (and newer) deps are typically needed).

csomh commented:
This PR was updated to include a proposal for `handle.paths`, clarified some defaults, renamed `package_paths` to `package.paths`, and updated examples accordingly.

csomh commented:
Let's keep this open until all talking points are settled with @lachmanfrantisek  and @praiskup.

praiskup commented:
> Let's keep this open until all talking points are settled with @lachmanfrantisek and @praiskup.

I'm fine here.  Thank you for the heads-up!

lachmanfrantisek commented on source-git-monorepos/README.md:
```diff
@@ -50,21 +89,367 @@ packages:
 - TODO Can archives stored in the dist-git repositories mapped to one or more
   paths in the upstream repo?
 - TODO Is support needed to have different packages generated from different
   git refs? (This would defy the point in having a monorepo, but some might
-  require it.)
+  require it, for example [Copr].)
```
E.g. Anaconda uses a different branch and packit config for each distribution -- and there are basically two sets of releases -- one made from a Fedora branch and the other one from the RHEL branch.
In that case, you have a different config in each branch so I am now not sure if this needs to be handled anyhow.
